### PR TITLE
[SPARK-19516][DOC] update public doc to use SparkSession instead of SparkContext

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

After Spark 2.0, `SparkSession` becomes the new entry point for Spark applications. We should update the public documents to reflect this.

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @sameeragarwal @hvanhovell

File: docs/programming-guide.md [hunk 1]
```diff
@@ -77,9 +76,9 @@ In addition, if you wish to access an HDFS cluster, you need to add a dependency
 Finally, you need to import some Spark classes into your program. Add the following lines:
 
 {% highlight scala %}
-import org.apache.spark.api.java.JavaSparkContext
-import org.apache.spark.api.java.JavaRDD
-import org.apache.spark.SparkConf
+import org.apache.spark.api.java.JavaSparkContext;
```
Comment (srowen) [inline docs/programming-guide.md]: You don't want semicolons in Scala right?

File: docs/programming-guide.md [hunk 2]
```diff
@@ -244,13 +239,13 @@ use IPython, set the `PYSPARK_DRIVER_PYTHON` variable to `ipython` when running
 $ PYSPARK_DRIVER_PYTHON=ipython ./bin/pyspark
 {% endhighlight %}
 
-To use the Jupyter notebook (previously known as the IPython notebook),
```
Comment (srowen) [inline docs/programming-guide.md]: Several extraneous whitespace changes but whatever

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/programming-guide.md (hunk 1 above)
Comment (cloud-fan) [inline docs/programming-guide.md]: hmm, shouldn't it be java code?

File: docs/programming-guide.md (hunk 2 above)
Comment (cloud-fan) [inline docs/programming-guide.md]: sorry my editor removes unnecessary whitespace automatically...

**Comment by cloud-fan** [comment]
several questions:

1. as `SparkSession` becomes the new entry point, now we require users to include spark-sql module in their sbt file or pom.xml, which seems weird if users only want to write a very simple spark application without SQL functionality. Shall we rename spark-sql to a more general name like `spark-structure`?
2. when it comes to java application with RDD, it's inconvenient to build `SparkSession` and then create `JavaSparkContext`, users can create `JavaSparkContext` directly.

cc @rxin @yhuai

**Comment by rxin** [comment]
I think the issue is that the programming guide should probably switch over to the DataFrame one as the primary one, and then the RDD one as a RDD programming guide.

cc @matei for his input.

File: docs/programming-guide.md [hunk 3]
```diff
@@ -76,10 +75,10 @@ In addition, if you wish to access an HDFS cluster, you need to add a dependency
 
 Finally, you need to import some Spark classes into your program. Add the following lines:
 
-{% highlight scala %}
-import org.apache.spark.api.java.JavaSparkContext
-import org.apache.spark.api.java.JavaRDD
-import org.apache.spark.SparkConf
+{% highlight java %}
```
Comment (srowen) [inline docs/programming-guide.md]: Ah right, this is the Java section. then yes the language should be fixed right here.

File: docs/programming-guide.md [hunk 4]
```diff
@@ -125,15 +124,12 @@ $ PYSPARK_PYTHON=/opt/pypy-2.5/bin/pypy bin/spark-submit examples/src/main/pytho
...
 <div data-lang="scala"  markdown="1">
 
-The first thing a Spark program must do is to create a [SparkContext](api/scala/index.html#org.apache.spark.SparkContext) object, which tells Spark
-how to access a cluster. To create a `SparkContext` you first need to build a [SparkConf](api/scala/index.html#org.apache.spark.SparkConf) object
-that contains information about your application.
-
-Only one SparkContext may be active per JVM.  You must `stop()` the active SparkContext before creating a new one.
+The first thing a Spark program must do is to create a [SparkSession](api/scala/index.html#org.apache.spark.sql.SparkSession) object, which tells Spark
+how to access a cluster. To create a `SparkSession`, you first need to call `SparkSession.builder` to get the builder, then set
+information about your application via the builder, and finall call `getOrCreate` on the builder to get the `SparkSession` instance.
```
Comment (heuermh) [inline docs/programming-guide.md]: `finall` &rarr; `finally`

File: docs/programming-guide.md [hunk 5]
```diff
@@ -153,13 +149,12 @@ JavaSparkContext sc = new JavaSparkContext(conf);
 
 <div data-lang="python"  markdown="1">
 
-The first thing a Spark program must do is to create a [SparkContext](api/python/pyspark.html#pyspark.SparkContext) object, which tells Spark
-how to access a cluster. To create a `SparkContext` you first need to build a [SparkConf](api/python/pyspark.html#pyspark.SparkConf) object
-that contains information about your application.
+The first thing a Spark program must do is to create a [SparkSession](api/python/pyspark.html#pyspark.sql.SparkSession) object, which tells Spark
+how to access a cluster. To create a `SparkSession`, you first need to call `SparkSession.builder` to get the builder, then set
+information about your application via the builder, and finall call `getOrCreate` on the builder to get the `SparkSession` instance.
```
Comment (heuermh) [inline docs/programming-guide.md]: `finall` &rarr; `finally`

File: docs/quick-start.md [hunk 6]
```diff
@@ -211,7 +199,7 @@ a cluster, as described in the [programming guide](programming-guide.html#initia
 It may seem silly to use Spark to explore and cache a 100-line text file. The interesting part is
 that these same functions can be used on very large data sets, even when they are striped across
 tens or hundreds of nodes. You can also do this interactively by connecting `bin/pyspark` to
-a cluster, as described in the [programming guide](programming-guide.html#initializing-spark).
+a cluster, as described in the [programming guide](rdd-programming-guide.html#initializing-spark).
```
Comment (cloud-fan) [inline docs/quick-start.md]: we probably wanna separate the previous programming guide into a basic programming guide and an RDD programming guide. The basic programming guide should talk about some basic knowledge like how to initialize spark, how to deploy spark, etc.

Comment (cloud-fan) [inline docs/quick-start.md]: cc @rxin

File: docs/quick-start.md [hunk 7]
```diff
@@ -10,12 +10,13 @@ description: Quick start tutorial for Spark SPARK_VERSION_SHORT
 This tutorial provides a quick introduction to using Spark. We will first introduce the API through Spark's
 interactive shell (in Python or Scala),
 then show how to write applications in Java, Scala, and Python.
-See the [programming guide](programming-guide.html) for a more complete reference.
 
 To follow along with this guide, first download a packaged release of Spark from the
 [Spark website](http://spark.apache.org/downloads.html). Since we won't be using HDFS,
 you can download a package for any version of Hadoop.
 
+Note that, before Spark 2.0, the main programming interface of Spark was Resilient Distributed Dataset (RDD). After Spark 2.0, RDD is replaced by Dataset, which is strongly typed like RDD, but with richer optimizations under the hood. The RDD interface is still supported, and you can get a more complete reference at [RDD programming guide](rdd-programming-guide.html). However, we highly recommend you to switch to use Dataset, which has better performance than RDD. See [SQL programming guide](sql-programming-guide.html) to get more information about Dataset.
```
Comment (srowen) [inline docs/quick-start.md]: Not a big deal, but it should probably say things like "_the_ Resilient Distributed Dataset" and "RDDs are replaced" vs "RDD is replaced" and "like an RDD" vs "like RDD"? and "the SQL programming guide", "the RDD programming guide"

File: docs/quick-start.md [hunk 8]
```diff
@@ -29,28 +30,28 @@ or Python. Start it by running the following in the Spark directory:
...
+scala> textFile.count() // Number of items in this Dataset
 res0: Long = 126 // May be different from yours as README.md will change over time, similar to other outputs
 
-scala> textFile.first() // First item in this RDD
+scala> textFile.first() // First item in this Dataset
 res1: String = # Apache Spark
 {% endhighlight %}
 
-Now let's use a transformation. We will use the [`filter`](programming-guide.html#transformations) transformation to return a new RDD with a subset of the items in the file.
+Now let's transform this Dataset to a new one. We will call the `filter` to return a new Dataset with a subset of the items in the file.
```
Comment (srowen) [inline docs/quick-start.md]: Just "call `filter`"?

File: docs/quick-start.md [hunk 9]
```diff
@@ -65,41 +66,41 @@ res3: Long = 15
 
     ./bin/pyspark
 
-Spark's primary abstraction is a distributed collection of items called a Resilient Distributed Dataset (RDD). RDDs can be created from Hadoop InputFormats (such as HDFS files) or by transforming other RDDs. Let's make a new RDD from the text of the README file in the Spark source directory:
+Spark's primary abstraction is a distributed collection of items called a Dataset. Datasets can be created from Hadoop InputFormats (such as HDFS files) or by transforming other Datasets. Due to Python's dynamic nature, we don't need the Dataset to be strongly typed in Python. As a result, all Datasets in Python are Dataset[Row], and we call it `DataFrame` to be consistent with the data frame concept in Pandas and R. Let's make a new DataFrame from the text of the README file in the Spark source directory:
```
Comment (srowen) [inline docs/quick-start.md]: strongly typed -> strongly-typed
(Same above, maybe elsewhere)

File: docs/quick-start.md [hunk 10]
```diff
@@ -137,37 +138,24 @@ res6: Array[(String, Int)] = Array((means,1), (under,2), (this,3), (Because,1),
...
-...     if a > b:
-...         return a
-...     else:
-...         return b
-...
-
->>> textFile.map(lambda line: len(line.split())).reduce(max)
-15
-{% endhighlight %}
+This first maps a line to an integer value and alias it with "numWords", creating a new DataFrame. `agg` is called on that DataFrame to find the largest word count. The arguments to `select` and `agg` are both _[Column](api/python/index.html#pyspark.sql.Column)_, we can use `df.colName` to get a column from a DataFrame, we can also import pyspark.sql.functions, which provides a lot of convenient functions to build a new Column from an old one.
```
Comment (srowen) [inline docs/quick-start.md]: alias it with "numWords" -> aliases it as "numWords"
The clauses beginning ", we can" should be new sentences?

**Review by srowen — CHANGES_REQUESTED** [review]

File: docs/quick-start.md [hunk 11]
```diff
@@ -438,7 +412,7 @@ Lines with a: 46, Lines with b: 23
 # Where to Go from Here
 Congratulations on running your first Spark application!
 
-* For an in-depth overview of the API, start with the [Spark programming guide](programming-guide.html),
+* For an in-depth overview of the API, start with the [RDD programming guide](rdd-programming-guide.html),
```
Comment (ueshin) [inline docs/quick-start.md]: Should this link to sql-programming-guide.html ?

Comment (sameeragarwal) [inline docs/quick-start.md]: and/or maybe just the API docs?

File: docs/quick-start.md [hunk 12]
```diff
@@ -438,8 +412,7 @@ Lines with a: 46, Lines with b: 23
 # Where to Go from Here
 Congratulations on running your first Spark application!
 
-* For an in-depth overview of the API, start with the [Spark programming guide](programming-guide.html),
-  or see "Programming Guides" menu for other components.
+* For an in-depth overview of the API, start with the [RDD programming guide](rdd-programming-guide.html) and the [SQL programming guide](sql-programming-guide.html), or see "Programming Guides" menu for other components.
```
Comment (cloud-fan) [inline docs/quick-start.md]: users still need to read RDD programming guide to learn some basic concepts like driver, executor, etc.

**Review by sameeragarwal — APPROVED** [review]
LGTM

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Comment by tdas** [comment]
@cloud-fan @sameeragarwal There are quite a few links in rest of Spark docs broken because of the name change from programming-guide -> rdd-programming-guide name change.

**Comment by cloud-fan** [comment]
ah, sorry for that, let me do a grep and fix them.

**Comment by tdas** [comment]
Don't worry. I am fixing it. I am having a PR soon.

**Comment by tdas** [comment]
My PR to fix them https://github.com/apache/spark/pull/18485
